BuddySave/Core/GamingSession.cs

```diff
@@ -1,57 +1,27 @@
-﻿using System.Diagnostics;
-using BuddySave.Core.Models;
-using BuddySave.System;
-using Microsoft.Extensions.Logging;
+﻿using BuddySave.Core.Models;
 
 namespace BuddySave.Core;
 
-public class GamingSession(
-    ILogger<GamingSession> logger,
-    ISharedSaveOrchestrator sharedSaveOrchestrator,
-    IProcessProvider processProvider)
-    : IGamingSession
+public class GamingSession(ILockManager lockManager, IServerSession serverSession, IClientSession clientSession) : IGamingSession
 {
-    public async Task RunServerWithAutoSave(GameSave gameSave, Session session, ServerParameters serverParameters)
-    {
-        if (string.IsNullOrWhiteSpace(serverParameters.Path))
-        {
-            throw new ArgumentException("No server path provided. Cannot start a gaming session.");
-        }
+	public async Task Play(GameSave gameSave, Session session, ServerParameters serverParameters, ClientParameters clientParameters)
+	{
+		var lockExists = lockManager.LockExists(gameSave);
+		var sessionToConnectTo = lockExists ? await lockManager.GetLockedSession(gameSave) : GetLocalSession(session);
 
-        var loadResult = await sharedSaveOrchestrator.Load(gameSave, session);
-        if (loadResult is not OrchestratorResult.Loaded)
-        {
-            return;
-        }
-        
-        var process = StartServer(serverParameters);
-        await WaitForServerToStop(process);
-        await sharedSaveOrchestrator.Save(gameSave, session);
-    }
+		var serverTask = Task.CompletedTask;
+		if (!lockExists)
+		{
+			serverTask = serverSession.RunServerWithAutoSave(gameSave, session, serverParameters);
+		}
 
-    private Process StartServer(ServerParameters serverParameters)
-    {
-        var workingDirectory = Path.GetDirectoryName(serverParameters.Path);
-        var startInfo = new ProcessStartInfo
-        {
-            FileName = serverParameters.Path,
-            Arguments = serverParameters.Arguments,
-            WorkingDirectory = workingDirectory,
-            UseShellExecute = true
-        };
+		clientSession.StartClient(sessionToConnectTo, clientParameters);
 
-        var process = processProvider.Start(startInfo);
-        var serverString = string.IsNullOrEmpty(serverParameters.Arguments)
-            ? serverParameters.Path
-            : $"{serverParameters.Path} {serverParameters.Arguments}";
-        logger.LogInformation(@$"Server started, waiting for exit: ""{serverString}""");
+		await serverTask;
```

Comment:
Wouldn't it be better to start awaiting for `serverTask` before starting client?

Reply:
Awaiting will be a blocking call. I guess I could do it on another thread.

From my experience the game client takes long enough to launch that the server starts up no problem :D. Would be interesting to check on a faster PC.